tree/30_generic_methods/command_execution_once.cf

```diff
@@ -0,0 +1,136 @@
+#####################################################################################
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, Version 3.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#####################################################################################
+
+# @name Command execution once
+# @description Execute a command only once on a node
+# @action
+# @documentation Execute a command only once.
+#
+# This method is useful for specific commands that should only be executed once per node.
+#
+# If you can spot a condition for the command execution by testing the state of its target,
+# it is better to use `condition_from_command` to test the state and `command_execution_result`
+# to run the command if necessary.
+#
+# When you use this method, it will:
+#
+# * If a `command_execution_once` has already been executed on this machine with the same
+#   unique id, define a `command_execution_once_${command}_kept` condition and do nothing
+# * If it is the first occurrence, it will execute the command and:
+#   * If the until parameter is `any`, it will consider the command as executed on the machine and define
+#     `command_execution_once_${command}_repaired/error` depending if the return code is in `ok_codes`
+#   * If the until parameter is `ok`, it will store consider it done only if the return code is in the
+#     `ok_codes` list and define `command_execution_once_${command}_repaired.` If not, it will retry at
+#     the next agent run and define `command_execution_once_${command}_error`
+#
+# If an exit code is not in the list it will lead to an error status. 
+# If you want 0 to be a success you have to list it in the ok_codes list
+#
+# #### Example
+#
+# If you use:
+#
+# ```
+# command_execution_once("command -a -t", "0", "ok", "my_program_setup")
+# ```
+#
+# Will retry to run `command -a -t` until it returns `0`. Then it will
+# not execute it again.
+#
+# @parameter command        The command to run
+# @parameter ok_codes       List of codes that produce a repaired status separated with commas (ex: 1,2,5). Defaults to 0.
+# @parameter_constraint ok_codes "allow_empty_string" : true
+# @parameter until          Try to execute the command until a particular state: 'ok', 'any' (defaults to 'any')
+# @parameter_constraint until "allow_empty_string" : true
+# @parameter_constraint until "select" : [ "", "any", "ok" ]
+# @parameter unique_id      To identify the command (allows to change the command without losing track)
```

Comment:
shouldn't it be optionnal, and default to command ?

Reply:
would be better indeed!